File: src/types/payload.ts
/**
 * 载荷生成模块类型定义
 */

// 生成模式 (向后兼容，保留旧值)
export type PayloadMode = 
  | 'simple' 
  | 'advanced' 
  | 'file_based' 
  | 'memory_shell' 
  | 'suo5_only'
  | 'file_common'      // 文件通用模式
  | 'file_proxy'      // 文件代理模式
  | 'file_hybrid'     // 文件混合模式

// 脚本类型
export type ScriptType = 'php' | 'jsp' | 'aspx' | 'asp'

// 功能类型
export type FunctionType = 'basic' | 'file_manager' | 'process_manager' | 'registry' | 'network'

// 注入类型
export type InjectionType = 'tomcat_filter' | 'spring_interceptor' | 'iis_httpmodule'

// 编码器类型
export type EncodeType = 
  | 'none' 
  | 'base64' 
  | 'xor' 
  | 'gzinflate' 
  | 'hex' 
  | 'urlencode' 
  | 'rot13'

// 加密算法
export type EncryptAlgo = 'aes128_cbc' | 'aes256_cbc' | 'aes256_gcm' | 'xor'

// 混淆级别
export type ObfuscationLevel = 'low' | 'medium' | 'high'

/**
 * Suo5 配置类型
 */
export interface Suo5Config {
  auth: string
  path: string
  timeout: number
}

/**
 * 载荷生成配置
 */
export interface PayloadConfig {
  mode: PayloadMode
  script_type: ScriptType
  function_type: FunctionType
  password: string
  encode_type?: EncodeType
  encrypt_algo?: EncryptAlgo
  obfuscation_level: ObfuscationLevel
  output_filename?: string
  template_name?: string
  injection_type?: InjectionType
  suo5_config?: Suo5Config
  self_destruct: boolean
}

/**
 * 客户端配置 (仅 Advanced 模式)
 */
export interface ClientConfig {
  key: string
  iv: string
  algorithm: string
  options: Record<string, any>
}

/**
 * 载荷生成结果
 */
export interface PayloadResult {
  code: string
  client_config: ClientConfig | null
  filename: string
  size: number
  success: boolean
  message?: string
}

/**
 * 编码器选项
 */
export const ENCODE_OPTIONS: Array<{ label: string; value: EncodeType }> = [
  { label: '无编码', value: 'none' },
  { label: 'Base64', value: 'base64' },
  { label: 'XOR', value: 'xor' },
  { label: 'GZInflate', value: 'gzinflate' },
  { label: 'Hex', value: 'hex' },
  { label: 'URL Encode', value: 'urlencode' },
  { label: 'ROT13', value: 'rot13' },
]

/**
 * 加密算法选项
 */
export const ENCRYPT_OPTIONS: Array<{ label: string; value: EncryptAlgo }> = [
  { label: 'AES-256-GCM', value: 'aes256_gcm' },
]

/**
 * 脚本类型选项
 */
export const SCRIPT_TYPE_OPTIONS: Array<{ label: string; value: ScriptType }> = [
  { label: 'PHP', value: 'php' },
  { label: 'JSP', value: 'jsp' },
  { label: 'ASPX', value: 'aspx' },
  { label: 'ASP', value: 'asp' },
]

/**
 * 功能类型选项
 */
export const FUNCTION_TYPE_OPTIONS: Array<{ label: string; value: FunctionType }> = [
  { label: '基础连接', value: 'basic' },
  { label: '文件管理', value: 'file_manager' },
  { label: '进程管理', value: 'process_manager' },
  { label: '注册表操作', value: 'registry' },
  { label: '网络操作', value: 'network' },
]

/**
 * 混淆级别选项
 */
export const OBFUSCATION_OPTIONS: Array<{ label: string; value: ObfuscationLevel }> = [
  { label: '低', value: 'low' },
  { label: '中', value: 'medium' },
  { label: '高', value: 'high' },
]

/**
 * 生成模式选项 (包含新旧模式，向后兼容)
 */
export const PAYLOAD_MODE_OPTIONS: Array<{ label: string; value: PayloadMode }> = [
  { label: '简单模式', value: 'simple' },
  { label: '高级模式', value: 'advanced' },
  { label: '文件落地', value: 'file_based' },
  { label: '内存 Shell', value: 'memory_shell' },
  { label: '仅 Suo5', value: 'suo5_only' },
  { label: '文件通用模式', value: 'file_common' },
  { label: '文件代理模式', value: 'file_proxy' },
  { label: '文件混合模式', value: 'file_hybrid' },
]

/**
 * 注入类型选项
 */
export const INJECTION_TYPE_OPTIONS: Array<{ label: string; value: InjectionType }> = [
  { label: 'Tomcat Filter', value: 'tomcat_filter' },
  { label: 'Spring Interceptor', value: 'spring_interceptor' },
  { label: 'IIS HTTP Module', value: 'iis_httpmodule' },
]

File: src-tauri/src/types/payload.rs
/// 载荷生成模块的类型定义
/// 用于前端和后端之间的 IPC 通信

use serde::{Deserialize, Serialize};

/// 载荷生成模式
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum PayloadMode {
    /// 极简一句话模式 - 仅编码/混淆 (向后兼容)
    Simple,
    /// 高级加密壳模式 - 强加密 (向后兼容)
    Advanced,
    /// 基于文件的载荷模式
    FileBased,
    /// 内存Shell模式
    MemoryShell,
    /// 仅Suo5模式
    Suo5Only,
}

/// 脚本类型
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum ScriptType {
    Php,
    Jsp,
    Aspx,
    Asp,
}

impl std::fmt::Display for ScriptType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ScriptType::Php => write!(f, "php"),
            ScriptType::Jsp => write!(f, "jsp"),
            ScriptType::Aspx => write!(f, "aspx"),
            ScriptType::Asp => write!(f, "asp"),
        }
    }
}

/// 注入类型
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum InjectionType {
    TomcatFilter,
    SpringInterceptor,
    IisHttpModule,
}

/// Suo5 配置结构体
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Suo5Config {
    pub auth: String,
    pub path: String,
    pub timeout: u32,
}

/// 编码器类型 (仅 Simple 模式使用)
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum EncodeType {
    None,
    Base64,
    Xor,
    GzInflate,
    Hex,
    UrlEncode,
    Rot13,
}

/// 加密算法 (仅 Advanced 模式使用)
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum EncryptAlgo {
    Aes128Cbc,
    Aes256Cbc,
    Xor,
}

/// 功能类型
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum FunctionType {
    Basic,
    FileManager,
    ProcessManager,
    Registry,
    Network,
}

/// 混淆级别
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum ObfuscationLevel {
    Low,
    Medium,
    High,
}

/// 载荷生成配置
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PayloadConfig {
    /// 生成模式
    pub mode: PayloadMode,
    /// 脚本类型
    pub script_type: ScriptType,
    /// 功能类型
    pub function_type: FunctionType,
    /// 连接密码
    pub password: String,
    /// 编码器类型 (仅 Simple 模式)
    pub encode_type: Option<EncodeType>,
    /// 加密算法 (仅 Advanced 模式)
    pub encrypt_algo: Option<EncryptAlgo>,
    /// 混淆级别
    pub obfuscation_level: ObfuscationLevel,
    /// 输出文件名 (可选)
    pub output_filename: Option<String>,
    /// 模板名称 (可选，用于自定义模板)
    pub template_name: Option<String>,
    /// 注入类型 (可选)
    pub injection_type: Option<InjectionType>,
    /// Suo5 配置 (可选)
    pub suo5_config: Option<Suo5Config>,
}

/// 客户端配置 (仅 Advanced 模式需要)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ClientConfig {
    /// 加密密钥
    pub key: String,
    /// 初始化向量 (IV)
    pub iv: String,
    /// 加密算法
    pub algorithm: String,
    /// 其他配置项
    pub options: serde_json::Value,
}

/// 载荷生成结果
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PayloadResult {
    /// 生成的源代码
    pub code: String,
    /// 客户端配置 (仅 Advanced 模式)
    pub client_config: Option<ClientConfig>,
    /// 文件名
    pub filename: String,
    /// 文件大小 (字节)
    pub size: u64,
    /// 成功标志
    pub success: bool,
    /// 消息
    pub message: Option<String>,
}

/// 载荷模板结构体
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PayloadTemplate {
    /// 模板名称
    pub name: String,
    /// 脚本类型
    pub script_type: ScriptType,
    /// 功能类型
    pub function_type: FunctionType,
    /// 模板代码
    pub code: String,
    /// 描述
    pub description: String,
    /// 创建时间
    pub created_at: String,
    /// 更新时间
    pub updated_at: String,
}

/// 错误类型
#[derive(Debug, thiserror::Error)]
pub enum GeneratorError {
    #[error("生成失败：{0}")]
    GenerationFailed(String),
    
    #[error("无效的密码：{0}")]
    InvalidPassword(String),
    
    #[error("文件操作失败：{0}")]
    IoError(#[from] std::io::Error),
    
    #[error("JSON 序列化失败：{0}")]
    JsonError(#[from] serde_json::Error),
    
    #[error("模板不存在：{0}")]
    TemplateNotFound(String),
    
    #[error("模板已存在：{0}")]
    TemplateExists(String),
}

pub type Result<T> = std::result::Result<T, GeneratorError>;
